application/apps/indexer/processor/src/text_source.rs

```diff
@@ -15,6 +15,46 @@ use tokio_util::sync::CancellationToken;
 const REDUX_READER_CAPACITY: usize = 1024 * 32;
 const REDUX_MIN_BUFFER_SPACE: usize = 10 * 1024;
 
+/// Convert byte's slice to UTF8 string.
+/// Checks for invalid characters and replace it with given char
+///
+/// # Arguments
+///
+/// * `bytes` - A byte's slice to be decoded.
+/// * `placeholder` - Char to replace with invalid UTF8 characters
+///
+/// # Returns
+///
+/// Returns valid UTF8 `String`.
+///
+fn bytes_to_utf8(bytes: &[u8], placeholder: Option<char>) -> String {
```

Comment:
that looks very complicated for something that should be part of the stdlib...and indeed there is an easier way:
``` Rust
String::from_utf8_lossy(bytes).to_string()
```
that will not allow us do specify our own replacement characters, but should be sufficient!
Bonus: will pass all your tests ;)

example:
``` rust
    #[test]
    fn test_invalid_utf8_easy() {
        let bytes = vec![0x68, 0x65, 0x6C, 0x6C, 0x6F, 0xC1, 0x80];
        assert!(
            std::str::from_utf8(&bytes).is_err(),
            "Should be invalid UTF-8"
        );
        let serialized = bytes_to_utf8(&bytes, Some('?'));
        assert_eq!(serialized, "hello��");
    }
```

Reply:
Thanks for it. Have no idea how I've missed it...With it indeed fix became trivial.